src/libcore/yk_swt_impl.c

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <err.h>
+#include <stdbool.h>
+
+struct mir_loc {
+    uint64_t crate_hash;
+    uint32_t def_idx;
+    uint32_t bb_idx;
+};
+
+#define TL_TRACE_INIT_CAP   1024
+
+static __thread struct mir_loc *tl_trace = NULL;
+static __thread uint32_t tl_trace_len = 0;
+static __thread uint32_t tl_trace_cap = 0;
+static __thread bool tl_tracing_en = false;
+
+// Start tracing on the current thread.
+// A new trace buffer is allocated and MIR locations are written into it. The
+// current thread must not already be tracing.
+void
+yk_swt_start_tracing_impl() {
+    if (tl_tracing_en) {
+        errx(EXIT_FAILURE, "%s: thread already tracing", __func__);
+    }
+
+    tl_trace = calloc(TL_TRACE_INIT_CAP, sizeof(struct mir_loc));
+    if (tl_trace == NULL) {
+        err(EXIT_FAILURE, "%s: calloc: ", __func__);
+    }
+
+    tl_trace_cap = TL_TRACE_INIT_CAP;
+    tl_tracing_en = true;
+}
+
+// Record a location into the trace buffer if tracing is enabled on the current thread.
+void
+yk_swt_rec_loc_impl(uint64_t crate_hash, uint32_t def_idx, uint32_t bb_idx)
+{
+    // If tracing is not currently active, then we do nothing.
+    if (!tl_tracing_en) {
+        return;
+    }
+
+    // Check if we need more space and reallocate if necessary.
+    if (tl_trace_len == tl_trace_cap) {
+        if (tl_trace_cap >= UINT32_MAX / 2) {
+            errx(EXIT_FAILURE, "%s: trace capacity would overflow", __func__);
+        }
+        size_t new_cap = tl_trace_cap * 2;
```

Comment:
This is a bit dangerous, I think: it means we're logarithmically increasing the buffer size. In other words, the bigger the buffer is, the more likely that we end up wasting lots of memory. We probably want some sort of threshold here that says something like "if the buffer is bigger than XMiB, then only add 1MiB extra storage" or similar.